### DISCUSSION TITLE
cURL using p12 certificate  for C++ client

### DISCUSSION THREAD

**Discussion by pkcs11newbie** [curl/curl#10664] [Q&A]
Hello,

I'm researching on latest cURL to connect C++ client to backend https server. 

FYI, I have generated .p12, containing certs (host and ca) and private key, protected by  a password.
%OPENSSL_CMD% pkcs12 -export -in XXXX.pem -inkey XXXX.privkey.pem -out XXXX.p12 -passout pass:aaaa -name XXX -certfile YYYY-ca.pem

Looking through some of SO [answers,](https://stackoverflow.com/questions/70604743/php-curl-request-for-a-p12-pfx-certificate-with-an-export-password-passphrase) 

If I were to do the following: 
curl_setopt($curl, CURLOPT_SSLCERTTYPE, "P12");
curl_setopt($curl, CURLOPT_SSLCERT, 'XXXX.p12');
curl_setopt($curl, CURLOPT_SSLCERTPASSWD, 'aaaa '); // 

I was wondering if this alone will successfully connect the client to the backend https server.

or 

Do I have to look into  [example](https://curl.se/libcurl/c/cacertinmem.html). If I'm not mistaken, implies to me that I can use OPENSSL to extract out host , private key and cacert and store in the sslcontext. cCURL, I suppose, then uses this context to connect to backend server. 

Any help, guidance greatly appreciated !!

Thanks.

**Comment by pkcs11newbie** [comment]
I'm running into similar issue mentioned https://github.com/curl/curl/issues/8966. Would appreciate insight into resolution. 
Was OpenSSL 3.0.X used to generate the pkcs12 file and tls connection to server was successful ? or converting pkcs ?

Thanks

**Reply by jay** [reply]
> I'm running into similar issue mentioned #8966. Would appreciate insight into resolution.

See https://stackoverflow.com/a/72600724

Also, I don't understand your original question.

**Reply by pkcs11newbie** [reply]
Hello,

Thank you for your response.

My original question was if libCurl can work with P12, given that p12 file contains host certificate, private key and CA chain certs without having to use openssl API to extract the pieces from p12. 

With just p12 tls handshake failed. I had to provide p12 and cafile as shown below for tls connection to be successful.
	curl_easy_setopt(curlHandle, CURLOPT_SSLCERTTYPE, "P12" );
	curl_easy_setopt(curlHandle, CURLOPT_SSLCERT, ***p12**** );
	curl_easy_setopt(curlHandle, CURLOPT_SSLCERTPASSWD, ********* );
       curl_easy_setopt( curlHandle, CURLOPT_CAINFO, "***cafile****" );

As far as the other question, my p12 was based of legacy algorithms and libCurl based of OpenSSL3.0 was having issues because of it. We'll just move to creating p12 with Openssl3.0, libCurl had no issue with that one.

Thanks
